File: Gold이상문제_정리/색종이붙이기_17136.java
package Gold이상문제_정리;

import java.util.*;
import java.io.*;

public class 색종이붙이기_17136 {
	static int[][] map;
	static int[] paper = new int[6];
	static int used = 987654321;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;

		Arrays.fill(paper, 5);
		map = new int[10][10];
		for (int i = 0; i < 10; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < 10; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		dfs(new Dot(0, 0), 0);
		if (used == 987654321)
			used = -1;
		System.out.println(used);
	}

	private static void dfs(Dot dot, int count)
	{
		// 끝에 도달하면
		if (dot.y >= 9 && dot.x > 9) {
			used = Math.min(used, count);
			return;
		}

		// 조건에 안맞으면
		if (used <= count) return;

		// 다음 행
		if (dot.x > 9) {
			dfs(new Dot(dot.y + 1, 0), count);
			return;
		}

		// 색종이있으면
		if (map[dot.y][dot.x] == 1) {
			// 5가지의 색종이에 대해서
			for (int size = 5; size >= 1; size--) {
				// 아직 종이가 있고, 붙일수 있다면
				if (paper[size] > 0 && canAttach(dot, size)) {
					// 붙여보고, 색종이 개수 줄여보고, 남은 칸수 줄인후 다음거 체크
					attach(dot, size, 2);
					paper[size]--;
					dfs(new Dot(dot.y, dot.x + 1), count+1);
					// 다시 돌아오면 원상복구후 진행
					attach(dot, size, 1);
					paper[size]++;
				}
			}
		}// 없으면 다음 칸
		else
			dfs(new Dot(dot.y, dot.x+1), count);
	}

	private static void attach(Dot dot, int size, int k) {
		for (int i = dot.y; i < dot.y + size; i++) {
			for (int j = dot.x; j < dot.x + size; j++) {
				map[i][j] = k;
			}
		}
	}

	private static boolean canAttach(Dot dot, int size) {
		// 문제있는 경우
		for (int i = dot.y; i < dot.y + size; i++) {
			for (int j = dot.x; j < dot.x + size; j++) {
				if (i >= 10 || j >= 10)
					return false;
				if (map[i][j] != 1) {
					return false;
				}
			}
		}
		// 문제없다면
		return true;
	}

	private static class Dot
	{
		int y;
		int x;

		public Dot(int y, int x) {
			this.y = y;
			this.x = x;
		}
	}
}

File: 코틀린_코테스터디/week13/색종이붙이기17136.kt
package 코틀린_코테스터디.week13

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer
import kotlin.math.min

/*
0 0 0 0 0 0 0 0 0 0
0 1 1 0 0 0 0 0 0 0
0 1 1 1 0 0 0 0 0 0
0 0 1 1 1 1 1 0 0 0
0 0 0 1 1 1 1 0 0 0
0 0 0 0 1 1 1 0 0 0
0 0 1 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
 */
class Problem17136 {
    companion object {
        private const val BLANK = 0
        private const val ONE = 1
        private const val STICKER = 2
    }

    private val map = Array(11) { IntArray(11) }
    private var minCount = Int.MAX_VALUE
    private val pages = mutableMapOf(1 to 5, 2 to 5, 3 to 5, 4 to 5, 5 to 5)

    fun solution() {
        val br = BufferedReader(InputStreamReader(System.`in`))

        for (y in 0 until 10) {
            val st = StringTokenizer(br.readLine())
            for (x in 0 until 10) {
                map[y][x] = st.nextToken().toInt()
            }
        }

        recursive(0, 0, 0)
        if (minCount == Int.MAX_VALUE) print(-1)
        else print(minCount)
    }

    private fun recursive(y: Int, x: Int, pageCount: Int) {
        if (y == 9 && x == 10) {
            minCount = min(minCount, pageCount)
            return
        }

        if (minCount <= pageCount) return

        if (x == 10) {
            recursive(y + 1, 0, pageCount)
            return
        }

        if (map[y][x] == ONE) {
            for (size in 5 downTo 1) {
                if (canCover(y, x, size)) {
                    cover(y, x, size, STICKER)
                    pages[size] = pages[size]!! - 1
                    recursive(y, x + 1, pageCount + 1)
                    cover(y, x, size, ONE)
                    pages[size] = pages[size]!! + 1
                }
            }
        } else recursive(y, x + 1, pageCount)
    }

    private fun canCover(y: Int, x: Int, size: Int): Boolean {
        if (y + size - 1 >= 10) return false
        if (x + size - 1 >= 10) return false
        if (pages[size]!! <= 0) return false

        for (ny in y until y + size) {
            for (nx in x until x + size) {
                if (map[ny][nx] != ONE) return false
            }
        }

        return true
    }

    private fun cover(y: Int, x: Int, size: Int, cov: Int) {
        for (ny in y until y + size) {
            for (nx in x until x + size) {
                map[ny][nx] = cov
            }
        }
    }
}

fun main() {
    Problem17136().solution()
}